DataFormats/Portable/interface/PortableHostCollectionReadRules.h

```diff
@@ -0,0 +1,76 @@
+#ifndef DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+#define DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+
+#include <TGenericClassInfo.h>
+#include <TVirtualObject.h>
+
+#include "DataFormats/Portable/interface/PortableHostCollection.h"
+#include "FWCore/Utilities/interface/concatenate.h"
+#include "FWCore/Utilities/interface/stringize.h"
+
+// read function for PortableHostCollection, called for every event
+template <typename T>
+static void readPortableHostCollection_v1(char *target, TVirtualObject *from_buffer) {
+  // extract the actual types
+  using Collection = T;
+  using Layout = typename Collection::Layout;
+
+  // valid only for PortableHostCollection<T>
+  static_assert(std::is_same_v<Collection, PortableHostCollection<Layout>>);
+
+  // proxy for the object being read from file
+  struct OnFile {
+    Layout &layout_;
+  };
+
+  // address in memory of the buffer containing the object being read from file
+  char *address = static_cast<char *>(from_buffer->GetObject());
+  // offset of the "layout_" data member
+  static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");
+  // reference to the Layout object being read from file
+  OnFile onfile = {*(Layout *)(address + layout_offset)};
+
+  // pointer to the Collection object being constructed in memory
+  Collection *newObj = (Collection *)target;
+
+  // move the data from the on-file layout to the newly constructed object
+  Collection::ROOTReadStreamer(newObj, onfile.layout_);
+}
+
+// put set_PortableHostCollection_read_rules in the ROOT namespace to let it forward declare GenerateInitInstance
+namespace ROOT {
+
+  // set the read rules for PortableHostCollection<T>;
+  // this is called only once, when the dictionary is loaded.
+  template <typename T>
+  static bool set_PortableHostCollection_read_rules(std::string const &type) {
+    // forward declaration
+    TGenericClassInfo *GenerateInitInstance(T const *);
+
+    // build the read rules
+    std::vector<ROOT::Internal::TSchemaHelper> readrules(1);
+    ROOT::Internal::TSchemaHelper *rule = &readrules[0];
+    rule->fTarget = "buffer_,layout_,view_";
+    rule->fSourceClass = type;
+    rule->fSource = type + "::Layout layout_;";
+    rule->fCode = type + "::ROOTReadStreamer(newObj, onfile.layout_)";
+    rule->fVersion = "[1-]";
+    rule->fChecksum = "";
+    rule->fInclude = "";
+    rule->fEmbed = false;
+    rule->fFunctionPtr = reinterpret_cast<void *>(::readPortableHostCollection_v1<T>);
```

Comment:
using
```c++
reinterpret_cast<void *>(::readPortableHostCollection_v1<T>);
```
instead of
```c++
(void *)TFunc2void(::readPortableHostCollection<T>);
```
avoids the dependency on the ROOT internals, and (looking at the ROOT internals) avoids using a `union` to reinterpret the pointer type.

[According to CppReference](https://en.cppreference.com/w/cpp/language/reinterpret_cast) it is OK to use `reinterpret_cast` to cast a function pointer to and from `void*`:
> 8) On some implementations (in particular, on any POSIX compatible system as required by [`dlsym`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/dlsym.html)), a function pointer can be converted to `void*` or any other object pointer, or vice versa. If the implementation supports conversion in both directions, conversion to the original type yields the original value, otherwise the resulting pointer cannot be dereferenced or called safely.

Reply:
This might not work on all platforms... At least in 2009 it triggered warnings on some platforms.